roofit/roofitcore/src/TestStatistics/LikelihoodJob.cxx

```diff
@@ -0,0 +1,254 @@
+         break;
+      }
+      }
+   } else {
+      throw std::logic_error("LikelihoodJob::update_bool only implemented on worker processes.");
+   }
+}
+
+
+void LikelihoodJob::update_parameters() {
+   if (get_manager()->process_manager().is_master()) {
+      for (std::size_t ix = 0u; ix < static_cast<std::size_t>(_vars.getSize()); ++ix) {
+         bool valChanged = !_vars[ix].isIdentical(_saveVars[ix], kTRUE);
+         bool constChanged = (_vars[ix].isConstant() != _saveVars[ix].isConstant());
+
+         if (valChanged || constChanged) {
+            if (constChanged) {
+               ((RooRealVar *) &_saveVars[ix])->setConstant(_vars[ix].isConstant());
+            }
+            // TODO: Check with Wouter why he uses copyCache in MPFE; makes it very difficult to extend, because copyCache is protected (so must be friend). Moved setting value to if-block below.
+            //          _saveVars[ix].copyCache(&_vars[ix]);
+
+            // send message to queue (which will relay to workers)
+            RooAbsReal * rar_val = dynamic_cast<RooAbsReal *>(&_vars[ix]);
+            if (rar_val) {
+               Double_t val = rar_val->getVal();
+               dynamic_cast<RooRealVar *>(&_saveVars[ix])->setVal(val);
+               Bool_t isC = _vars[ix].isConstant();
+               update_real(ix, val, isC);
+            }
+         }
+      }
+   }
+}
+
+
+double LikelihoodJob::return_result() const {
+   return result;
+}
+
+void LikelihoodJob::evaluate() {
+   if (get_manager()->process_manager().is_master()) {
+      // update parameters that changed since last calculation (or creation if first time)
+      update_parameters();
+
+      // master fills queue with tasks
+      for (std::size_t ix = 0; ix < get_manager()->process_manager().N_workers(); ++ix) {
+         get_manager()->queue().add({id, ix});
+      }
+      N_tasks_at_workers = get_manager()->process_manager().N_workers();
+
+      // wait for task results back from workers to master
+      gather_worker_results();
+
+      // put the results in vectors for calling sum_of_kahan_sums (TODO: make a map-friendly sum_of_kahan_sums)
+//      std::vector<double> results_vec, carrys_vec;
+//      for (auto const &item : results) {
+//         results_vec.emplace_back(item.second);
+//         carrys_vec.emplace_back(carrys[item.first]);
+//      }
+//
+//      // sum task results
+//      std::tie(result, carry) = sum_of_kahan_sums(results_vec, carrys_vec);
+      std::tie(result, carry) = sum_of_kahan_sums(results, carrys);
+      apply_offsetting(result, carry);
+   }
+}
+
+// --- RESULT LOGISTICS ---
+
+//void LikelihoodJob::send_back_task_result_from_worker(std::size_t task) {
+//   get_manager()->messenger().send_from_worker_to_queue(id, task, result, carry);
+//}
+
+void LikelihoodJob::send_back_task_result_from_worker(std::size_t /*task*/) {
+   get_manager()->messenger().send_from_worker_to_master(result, carry);
```

Comment:
This all doesn't work, e.g. this doesn't send task ID at the moment, see LikelihoodGradientJob for a better implementation.